sentry-core/src/envelope.rs

```diff
@@ -0,0 +1,148 @@
+    }
+}
+
+/// A Sentry Envelope.
+///
+/// An Envelope is the data format that Sentry uses for Ingestion. It can contain
+/// multiple Items, some of which are related, such as Events, and Event Attachments.
+/// Other Items, such as Sessions are independant.
+///
+/// See the [documentation on Envelopes](https://develop.sentry.dev/sdk/envelopes/)
+/// for more details.
+#[derive(Default, Debug, PartialEq)]
+pub struct Envelope {
+    event_id: Option<Uuid>,
+    items: Vec<EnvelopeItem>,
+}
+
+impl Envelope {
+    /// Creates a new empty Envelope.
+    pub fn new() -> Envelope {
+        Default::default()
+    }
+
+    /// Returns the Envelopes Uuid, if any.
+    pub fn uuid(&self) -> Option<&Uuid> {
+        self.event_id.as_ref()
+    }
+
+    /// Returns the [`Event`] contained in this Envelope, if any.
+    ///
+    /// [`Event`]: protocol/struct.Event.html
+    pub fn event(&self) -> Option<&Event<'static>> {
+        // until we actually add more items:
+        #[allow(clippy::unnecessary_filter_map)]
+        self.items
+            .iter()
+            .filter_map(|item| match item {
+                EnvelopeItem::Event(event) => Some(event),
+            })
+            .next()
+    }
+
+    /// Serialize the Envelope into the given [`Write`].
+    ///
+    /// [`Write`]: https://doc.rust-lang.org/std/io/trait.Write.html
+    pub fn to_writer<W>(&self, mut writer: W) -> std::io::Result<()>
+    where
+        W: Write,
+    {
+        let mut item_buf = Vec::new();
+
+        // write the headers:
+        let event_id = self.uuid();
+        match event_id {
+            Some(uuid) => writeln!(writer, r#"{{"event_id":"{}"}}"#, uuid)?,
+            _ => writeln!(writer, "{{}}")?,
+        }
+
+        // write each item:
+        for item in &self.items {
+            // we write them to a temporary buffer first, since we need their length
+            serde_json::to_writer(
+                &mut item_buf,
+                match item {
+                    EnvelopeItem::Event(event) => event,
+                },
+            )?;
+            let item_type = match item {
+                EnvelopeItem::Event(_) => "event",
+            };
+            writeln!(
+                writer,
+                r#"{{"type":"{}","length":{}}}"#,
+                item_type,
+                item_buf.len()
+            )?;
+            writer.write_all(&item_buf)?;
+            writeln!(writer)?;
+            item_buf.clear();
+        }
+
+        Ok(())
+    }
+}
+
+impl From<Event<'static>> for Envelope {
+    fn from(event: Event<'static>) -> Self {
+        let mut envelope = Self::default();
+        envelope.event_id = Some(event.event_id);
+        envelope.items.push(event.into());
+        envelope
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    fn to_buf(envelope: Envelope) -> std::io::Result<Vec<u8>> {
+        let mut vec = vec![];
+        envelope.to_writer(&mut vec)?;
+        Ok(vec)
```

Comment:
Since writing to `Vec` is infallible you could remove the `Result` here and unwrap right away.  (Additionally you're just unwrapping this function at al callers anyway afaik)